app/ts/TradingUI/Trading.tsx

```diff
@@ -57,16 +58,59 @@ const TradingView = ({ maybeReadClient, maybeWriteClient, marketData, currentTim
 		}
 	})
 
-	const refresh = async () => {
+	const updateSharesOut = async (daiInput: bigint) => {
 		if (maybeWriteClient.deepValue === undefined) return
 		if (maybeReadClient.deepValue === undefined) return
 		if (marketData.deepValue === undefined) return
+		const baseSharesExpected = daiInput / marketData.deepValue.numTicks
+		const expectedSwapShares = await expectedSharesAfterSwap(maybeReadClient.deepValue, marketData.deepValue.marketAddress, yesSelected.value === 'No', baseSharesExpected)
+		expectedSharesOut.deepValue = baseSharesExpected + expectedSwapShares
+	}
+
+	useSignalEffect(() => {
+		if (maybeWriteClient.deepValue === undefined) return
+		if (maybeReadClient.deepValue === undefined) return
+		if (marketData.deepValue === undefined) return
+
+		if (buySelected.value === 'Buy') {
+			if (daiInputAmountToBuy.deepValue === undefined) return
+			// TODO, fix race conditions
+			updateSharesOut(daiInputAmountToBuy.deepValue)
+		} else {
+			if (yesSelected.value === 'Yes') {
+				console.error('TODO: not implemented')
+			} else {
+				console.error('TODO: not implemented')
+			}
+		}
+	})
+
+	const updateShareBalances = async () => {
+		if (maybeWriteClient.deepValue === undefined) return
+		if (marketData.deepValue === undefined) return
 		const shareBalances = await getShareBalances(maybeWriteClient.deepValue, marketData.deepValue.marketAddress, maybeWriteClient.deepValue.account.address)
 		invalidBalance.deepValue = shareBalances[0]
 		noBalance.deepValue = shareBalances[1]
 		yesBalance.deepValue = shareBalances[2]
+	}
+
+	useEffect(() => {
+		updateShareBalances()
+	}, [])
```

Comment:
Isn't this unnecessary now that you have the `useSignalEffect` above?

Reply:
the above useSignalEffect does not call updateShareBalances()

Reply:
but yeah, this is bit wrong and should be a signal effect